io-sim/src/Control/Monad/IOSim/Internal.hs

```diff
@@ -1046,17 +1046,17 @@ saveTVar TVar{tvarCurrent, tvarUndo} = do
 revertTVar :: TVar s a -> ST s ()
 revertTVar TVar{tvarCurrent, tvarUndo} = do
     -- pop the undo stack, and revert the current value
-    (v:vs) <- readSTRef tvarUndo
-    !_ <- writeSTRef tvarCurrent v
-    !_ <- writeSTRef tvarUndo    vs
+    vs <- readSTRef tvarUndo
+    !_ <- writeSTRef tvarCurrent (head vs)
+    !_ <- writeSTRef tvarUndo    (tail vs)
```

Comment:
Why this change? I actually prefer how it was previously. you can rename `(v:vs)` to `(head:tail)` and recover readability

Reply:
Also maybe these changes could go into a different commit?

Reply:
`ST` has no `MonadFail` instance (`ghc-9.4.2`)

Reply:
Oh I see!!